### DISCUSSION TITLE
Unable to add Windows Community Toolkit

### DISCUSSION THREAD

**Discussion by joer33304** [unoplatform/uno#8993] [Q&A] [answered]
I am trying to add the Windows Community Toolkit package. I manually added the Microsoft toolkit to the windows project and the Uno toolkit to wasm. The windows project compiles, but the wasm shows error:

```
Error    UNOB0001    Cannot build with both Uno.WinUI and Uno.UI nuget packages referenced.
```

Uno version is 4.3.6, toolkit 7.1.11

**Comment by jeromelaban** [comment, marked as answer]
Thanks for the report.

The documentation is not up to date for the use of WinUI, but when you reference `Uno.WinUI`, you'll need to use WinUI compatible packages. In the case of [Community Toolkit](https://github.com/unoplatform/Uno.WindowsCommunityToolkit), you'll need to use `Uno.CommunityToolkit.*` packages.

**Reply by jeromelaban** [reply]
This is the updated documentation: https://github.com/unoplatform/uno/blob/master/doc/articles/uno-community-toolkit.md

**Reply by joer33304** [reply]
Now I am confused. I was following the example at https://platform.uno/docs/articles/uno-community-toolkit.html?tabs=tabid-vswin which is using Uno.Microsoft.Toolkit.Uwp.UI.Controls.DataGrid NOT Uno.CommunityToolkit.*. The sample at https://github.com/unoplatform/Uno.Samples/tree/master/UI/UnoWCTDataGridSample uses the same reference. So what's the difference between Uno.CommunityToolkit.* and  Uno.Microsoft.Toolkit ?

The issue in my project referencing WinUI is the fact that I added an UNO platform library to the project which I referenced in the platform projects . The UNO platform library has a reference to Uno.WinUI and this then gets copied over to the projects.

**Reply by joer33304** [reply]
Ah just now saw your update. Let me read it and see if it solves it :)

**Reply by joer33304** [reply]
Ok so the updated docs helped, old error gone but now a new error appeared ONLY in the windows app:

`Error	CS0433	The type 'SuspendingEventArgs' exists in both 'Microsoft.Windows.SDK.NET, Version=10.0.22000.24, Culture=neutral, PublicKeyToken=31bf3856ad364e35' and 'Uno, Version=255.255.255.255, Culture=neutral, PublicKeyToken=null'	UNO_V1.Windows	`

**Reply by jeromelaban** [reply]
This likely means that you're still referencing an `Uno.CommunityToolkit` package, either explicitly or indirectly. Do you have an existing project that you are referencing?

**Reply by joer33304** [reply]
I did a global search in my whole solution folder. There is only one reference in my uno platform library project and this has the conditional reference, so it should not be added to the windows project. Below is the proj file

```
<Project Sdk="MSBuild.Sdk.Extras/3.0.23">
	<PropertyGroup>
		<TargetFrameworks>net6.0-windows10.0.19041;netstandard2.1;net6.0-ios;net6.0-macos;net6.0-maccatalyst;net6.0-android</TargetFrameworks>
		<!-- Ensures the .xr.xml files are generated in a proper layout folder -->
		<GenerateLibraryLayout>true</GenerateLibraryLayout>
	</PropertyGroup>
	<ItemGroup>
		<PackageReference Include="Uno.WinUI" Version="4.3.8" />
	</ItemGroup>
	<ItemGroup Condition="'$(TargetFramework)' != 'net6.0-windows10.0.19041'">
		<PackageReference Include="CommunityToolkit.WinUI.UI.Controls" Version="7.1.2" />
	</ItemGroup>
	<ItemGroup Condition="'$(TargetFramework)'=='net6.0-windows10.0.19041'">
		<PackageReference Include="Uno.Microsoft.Toolkit.Uwp.UI.Controls" Version="7.1.11" />
		<PackageReference Include="Microsoft.WindowsAppSDK" Version="1.0.0" />
		<PackageReference Include="Microsoft.Toolkit.Uwp.UI.Controls" Version="7.1.2" />
		<FrameworkReference Update="Microsoft.Windows.SDK.NET.Ref" RuntimeFrameworkVersion="10.0.22000.24" />
		<FrameworkReference Update="Microsoft.Windows.SDK.NET.Ref" TargetingPackVersion="10.0.22000.24" />
	</ItemGroup>
	<ItemGroup Condition="'$(TargetFramework)' != 'net6.0-windows10.0.19041'">
		<Page Include="**\*.xaml" Exclude="bin\**\*.xaml;obj\**\*.xaml" />
		<Compile Update="**\*.xaml.cs">
			<DependentUpon>%(Filename)</DependentUpon>
		</Compile>
	</ItemGroup>
	<ItemGroup>
		<UpToDateCheckInput Include="**\*.xaml" Exclude="bin\**\*.xaml;obj\**\*.xaml" />
	</ItemGroup>
</Project>
```

**Reply by jeromelaban** [reply]
Thanks. The community toolkit packages references lines are inverted in your project, see [this section](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-community-toolkit.md#tab/tabid-winui).

Here's a correct one:
```xml
	<ItemGroup Condition="'$(TargetFramework)' != 'net6.0-windows10.0.19041'">
		<PackageReference Include="Uno.CommunityToolkit.WinUI.UI.Controls" Version="7.1.2" />
	</ItemGroup>
	<ItemGroup Condition="'$(TargetFramework)'=='net6.0-windows10.0.19041'">
		<PackageReference Include="Microsoft.Toolkit.Uwp.UI.Controls" Version="7.1.11" />
		<PackageReference Include="Microsoft.WindowsAppSDK" Version="1.0.0" />
		<PackageReference Include="Microsoft.Toolkit.Uwp.UI.Controls" Version="7.1.2" />
		<FrameworkReference Update="Microsoft.Windows.SDK.NET.Ref" RuntimeFrameworkVersion="10.0.22000.24" />
		<FrameworkReference Update="Microsoft.Windows.SDK.NET.Ref" TargetingPackVersion="10.0.22000.24" />
	</ItemGroup>
```

**Reply by joer33304** [reply]
So 2 weeks and countless hours later I still don't get this to run. One issue/confusion is on what "windows" project is created. If I use the command line a windows.uwp app is added, here the Microsoft.Toolkit.Uwp.UI.Controls works fine. But when you use the Visual Studio template, it creates a windows app which adding the toolkit reference fails saying its not compatible.

I think it would be good if command templates and vs templates would be the same. So how many windows app types are there, what's the difference and what is correct, the template or vs ? The command template created UWP app also strangely has no "edit project" file missing when you right click the project.

Last but not least, what toolkit reference is valid for the vs template created windows app ?
Your .proj example above is incorrect, you have the Microsoft.Toolkit.Uwp.UI.Control twice with different versions.

**Reply by jeromelaban** [reply]
[These sections](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-community-toolkit.md#winui--winappsdk-2) are about the WinUI vs. UWP differences. `Uno.Microsoft.Toolkit.Uwp.*` is about UWP, `Uno.CommunityToolkit.WinUI.*` is about WinAppSDK/WinUI.

If you have specific errors, let us know.

**Reply by joer33304** [reply]
Ok , I think finally understood but now the problem is that the namespaces are different for the toolkits so XAML does not work. Unfortuntely the IsApiContractPresent seems to not have an "else if" functionality, so I get a duplicate declaration error using below.

 In code behind we have compiler directives, how to use the controls in the shared projects XAML file if the namespaces are different ?  

```
xmlns:controls="using:Community.Toolkit.Uwp.UI.Controls?IsApiContractPresent(Uno.WinUI)"
xmlns:controls="using:Microsoft.Toolkit.Uwp.UI.Controls?IsApiContractNotPresent(Uno.WinUI)"
```

**Reply by joer33304** [reply]
@jeromelaban any help on this final issue ?

**Reply by jeromelaban** [reply]
XAML does not support conditional compilation this way, both WinAppSDK and UWP will not compile with such namespace definitions.

Is your objective to support both WinUI and UWP?

**Reply by joer33304** [reply]
No I just want to support one. I have more overall luck with the VS templates instead of the dotnet templates and the VS creates a windows app. The issue is that all non windows projects (mobile etc) are using Microsoft.Toolkit.Uwp.UI.Controls so if I want to use the shared XAML (the idea of UNO or ?) how do I do this if I need two different namespaces for windows app and the others

**Reply by jeromelaban** [reply]
Good. As you'll only use WinUI, you need to use the references provided [by this section](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-community-toolkit.md#tab/tabid-winui). The Community Toolkit uno packages (`Uno.CommunityToolkit.WinUI.UI.Controls`) are compatible with the official ones (`CommunityToolkit.WinUI.UI.Controls`).

You should not need to use any of the `Microsoft.Toolkit.Uwp.UI` namespaces, as those are for UWP only.

**Reply by joer33304** [reply]
Ah what a relief it finally works ! So ONLY using CommunityToolkit for ALL projects did the trick :)

Only remaining issue is the intellisense error is showing up again, this time saying it's 
only unavailable for Android. This happens no matter which project I select in the top VS bar.
I already recompiled, closed VS, deleted the .vs folder but it's always appearing after compiling the solution.

![Screenshot_7](https://user-images.githubusercontent.com/2421310/175847738-bec86a9b-5dae-4ef1-9b16-cc308bb7f56b.png)

**Reply by jeromelaban** [reply]
This specific issue is curious. Are you able to run the `net6.0-android` target for the mobile project ?

**Reply by joer33304** [reply]
You mean if the android app runs ? Yes all runs perfectly fine, it's just an intellisense error. As you can see from the screenshot below, intellisense actually works, I think it's only red because of the android "not available". I have a "global" dll where I add all my references, do I have some wrong target setting there ? I am questioning if netstandard2.1 as the wasm is net6.

`		<TargetFrameworks>netstandard2.1;xamarinios10;MonoAndroid12.0;net6.0-windows10.0.19041.0</TargetFrameworks> `

![Screenshot_8](https://user-images.githubusercontent.com/2421310/175948325-ae2fae9c-0d77-4d27-bfaf-925554e0b5f7.png)
![Screenshot_9](https://user-images.githubusercontent.com/2421310/175948365-77a8ec64-33cb-445c-8024-a612e9f42a1e.png)

**Reply by jeromelaban** [reply]
For Uno, you will not be able to mix `net6.0-android` and `monoandroid12.0`. If you heads are `net6-XXX`, your dependencies must include `net6-XX` as well. You can use the `Uno Platform Library` template to begin with.

**Reply by joer33304** [reply]
Ok I updated all projects and referenced projects to only use the net6-xx versions. All compiles, all runs but still intellisense does not work. Is there any verbose log somewhere I can look at ?

![Screenshot_16](https://user-images.githubusercontent.com/2421310/177416380-dbe9c8a7-9a55-4ee9-afe6-f3d382293b57.png)
![Screenshot_17](https://user-images.githubusercontent.com/2421310/177416416-0d5a42e9-e8f4-466c-b1a6-905929010695.png)
![Screenshot_13](https://user-images.githubusercontent.com/2421310/177416452-d3cee6d3-1530-4e17-9d29-d1c8886e04c9.png)
![Screenshot_14](https://user-images.githubusercontent.com/2421310/177416524-a1830302-0ad0-47ad-bc16-d5eafcb81769.png)
